uxarray/_mpas.py

```diff
@@ -0,0 +1,161 @@
+            "long_name": "longitude of mesh nodes",
+            "units": "degrees_east",
+        })
+
+    out_ds['Mesh2_node_y'] = xr.DataArray(
+        latVertex,
+        dims=["nMesh2_node"],
+        attrs={
+            "standard_name": "latitude",
+            "long_name": "latitude of mesh nodes",
+            "units": "degrees_north",
+        })
+
+    # centers of primal mesh cells (in degrees)
+    lonCell = np.rad2deg(in_ds['lonCell'].values)
+    latCell = np.rad2deg(in_ds['latCell'].values)
+
+    out_ds['Mesh2_face_x'] = xr.DataArray(
+        lonCell,
+        dims=["nMesh2_face"],
+        attrs={
+            "standard_name": "longitude",
+            "long_name": "longitude of center nodes",
+            "units": "degrees_east",
+        })
+
+    out_ds['Mesh2_face_y'] = xr.DataArray(
+        latCell,
+        dims=["nMesh2_face"],
+        attrs={
+            "standard_name": "latitude",
+            "long_name": "latitude of center nodes",
+            "units": "degrees_north",
+        })
+
+    # vertex indices that surround each primal mesh cell
+    verticesOnCell = in_ds['verticesOnCell'].values
+
+    # max number of edges & vertices that make up each cell
+    nEdgesOnCell = in_ds['nEdgesOnCell'].values
+
+    # correct to be zero-indexed and to use fill values
+    verticesOnCell_fill = _add_fill_values(verticesOnCell, nEdgesOnCell)
+
+    out_ds["Mesh2_face_nodes"] = xr.DataArray(
+        data=verticesOnCell_fill,
+        dims=["nMesh2_face", "nMaxMesh2_face_nodes"],
+        attrs={
+            "cf_role": "face_node_connectivity",
+            "_FillValue": fill_val,
+            "start_index": int_dtype(0)
+        })
+
+    # add Mesh2_edge_nodes if available
+    if "verticesOnEdge" in in_ds:
+        # vertex indices that saddle a given edge
+        verticesOnEdge = in_ds['verticesOnEdge'].values
+
+        # convert to zero-indexed
+        verticesOnEdge -= 1
+
+        out_ds["Mesh2_edge_nodes"] = xr.DataArray(
+            data=verticesOnEdge,
+            dims=["nMesh2_edge", "Two"],
+            attrs={
+                "cf_role": "edge_node_connectivity",
+                "start_index": int_dtype(0)
+            })
+
+
+def _add_fill_values(verticesOnCell, nEdgesOnCell):
+    """Corrects the input (``verticesOnCell``) to be zero-indexed and to use
+    fill-values instead of padding / zero-fills.
+
+    Parameters
+    ----------
+    verticesOnCell : np.ndarray
+        Vertex indices that surround a given cell
+
+    nEdgesOnCell : np.ndarray
+        Number of edges on a given cell
+    """
+    # convert to unsigned integers
+    verticesOnCell = verticesOnCell.astype(int_dtype)
+
+    # iterate over each primal mesh cell
+    for D, pad_start_idx in enumerate(nEdgesOnCell):
+        # convert non-padded values to zero-index
+        verticesOnCell[D, 0:pad_start_idx] -= 1
+        # replace remaining padding or zeros with fill_value
+        verticesOnCell[D, pad_start_idx:] = fill_val
```

Comment:
Iterating over cells in python is quite slow.  I have always iterated over the `maxEdges` dimension instead in loops like this one:
```suggestion
    maxEdges = verticesOnCell.shape[1]
    for vert_idx in range(maxEdges):
        mask = vert_idx < nEdgesOnCell
        # convert non-padded values to zero-index
        verticesOnCell[mask, vert_idx] -= 1
        # replace remaining padding or zeros with fill_value
        verticesOnCell[np.logical_not(mask), vert_idx] = fill_val
```

Reply:
I ran `pytest test` locally with these changes and it passes (not sure how rigorous a test that is).

I'm happy to do some other testing but not sure how much can be done with uxarray yet.  The docs don't seem to go much beyond reading in files and getting connectivity info in various ways. (I fully understand that this is a work in progress.)

Reply:
Now that I'm taking a closer look at this (it's been a long week), I'm noticing that within `verticesOnCell`, the shapes that have sides n < 7 are all padded out with the same value, which doesn't actually exist as a vertex index in the mesh file. While I was trying to test this on my own with Polymesh, I ran into downstream issues with numpy indexing for coordinate values (ie; `lonCell[verticesOnCell]`, `latCell[verticesOnCell]`).

I've done this differently within my own code by having each individual cell connect back to each individual "origin" vertex. It's a little clunky, but is marginally faster this way as well (~20-30 ms):

```
%%time
# MPAS indexing starts at 1, not 0 (any original 0 value ~ NaN)
face_nodes = c_mesh_ds.verticesOnCell.values - 1

# Replaces all -1 values with first value in array to force 7-sided shapes
face_nodes[:, 4] = np.where(face_nodes[:, 4] == -1, face_nodes[:, 0], face_nodes[:, 4])
face_nodes[:, 5] = np.where(face_nodes[:, 5] == -1, face_nodes[:, 0], face_nodes[:, 5])
face_nodes[:, 6] = np.where(face_nodes[:, 6] == -1, face_nodes[:, 0], face_nodes[:, 6])

# Ensures face nodes are returned as integers
face_nodes = face_nodes.astype(np.uint32)
```

```
CPU times: user 27.9 ms, sys: 32.2 ms, total: 60.1 ms
Wall time: 58.2 ms
```

Reply:
@cdeciampa, yes, that's another approach that's good for visualization (and I believe also the approach expected by the SCRIP format).  I don't know if that's an appropriate solution for uxarray more broadly.  That presumably needs to be a decision on what the standard uxarray format should be.

Reply:
@philipc2, my version accounts for all of that.  The mask array will be true for cells that have a vertex with the given index and will be false for cells that don't have it.  It is vastly more efficient to iterate over 7 to 9 vertices than it is to iterate over potentially tens of millions of cells.  So I stand by my version being much preferable, especially for very large MPAS meshes.  The result should be the same as your loop, as you can check.